modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheLockImpl2Unfair.java

```diff
@@ -0,0 +1,720 @@
+            final IgniteLogger log) {
+
+            assert nodeId != null;
+            assert key != null;
+            assert lockView != null;
+            assert listener != null;
+            assert ctx != null;
+            assert log != null;
+
+            this.key = key;
+            this.lockView = lockView;
+            this.listener = listener;
+            this.nodeId = nodeId;
+            this.log = log;
+
+            acquireProcessor = new AcquireUnfairProcessor(nodeId);
+            releaseProcessor = new ReleaseUnfairProcessor(nodeId);
+            lockIfFreeProcessor = new LockIfFreeUnfairProcessor(nodeId);
+            lockOrRemoveProcessor = new LockOrRemoveUnfairProcessor(nodeId);
+
+            acquireListener = new IgniteInClosure<IgniteInternalFuture<EntryProcessorResult<Boolean>>>() {
+                @Override public void apply(IgniteInternalFuture<EntryProcessorResult<Boolean>> future) {
+                    try {
+                        EntryProcessorResult<Boolean> result = future.get();
+
+                        if (result.get())
+                            listener.release();
+                    }
+                    catch (IgniteCheckedException e) {
+                        if (log.isDebugEnabled())
+                            log.debug("Acquire invoke has failed: " + e);
+
+                        listener.fail(U.convertException(e));
+                    }
+                }
+            };
+
+            releaseListener = new IgniteInClosure<IgniteInternalFuture<EntryProcessorResult<UUID>>>() {
+                @Override public void apply(IgniteInternalFuture<EntryProcessorResult<UUID>> future) {
+                    try {
+                        EntryProcessorResult<UUID> result = future.get();
+
+                        // invokeAsync return null if EntryProcessor return null too.
+                        if (result != null) {
+                            UUID nextNode = result.get();
+
+                            if (nextNode != null) {
+                                if (nodeId.equals(nextNode))
+                                    listener.release();
+                                else {
+                                    ctx.io().send(nextNode,
+                                        new GridCacheLockImpl2Unfair.ReleasedMessage(ctx.cacheId(), key.name()), P2P_POOL);
+                                }
+                            }
+                        }
+                    }
+                    catch (IgniteCheckedException e) {
+                        if (log.isDebugEnabled())
+                            log.debug("Release message sending has failed: " + e);
+                    }
+                }
+            };
+        }
+
+        /** */
+        private IgniteInternalFuture<EntryProcessorResult<Boolean>> tryAcquireOrAdd() {
+            return lockView.invokeAsync(key, acquireProcessor);
+        }
+
+        /** */
+        private boolean tryAcquire() {
```

Comment:
may be refactor pull member up and override in fair and unfair impls. This applicable for all methods that have equal names in both fait and unfair impl classes. Also javadocs )

Reply:
Implementations haven't any crossing, so such refactor just will add one useless interface.